Control interface for BatteryPool

shsms opened the pull request:
This PR updates the PowerDistributingActor to remove its multi-user interface, make its accessible only through the BatteryPool,  with these methods:

1. set_power :: send a PSC power request to the power distributor.
2. charge :: send a charge power (+ve) to the power distributor.
3. discharge :: send a discharge power (+ve) to the power distributor.
4. power_distribution_results :: get a receiver to power distributor responses.

shsms committed:
Use `abc.Set` in power distributing request objects

The PowerDistributingActor doesn't need to update request objects.
Some upstream typing has been updated to be read-only as well, to
support this.

This allows downstream users to also work with just `abc.Set` objects.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Update generated data-pipeline channel names to be more specific

These generated channel names are used to identify channels in the
channel registry.  This change allows to use the same channel registry
for other requirements as well, while also keeping the channels easy
to identify based on name.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Add BatteryPool methods for communicating with the PowerDistributor

1. set_power :: send a PSC power request to the power distributor.
2. charge :: send a charge power (+ve) to the power distributor.
3. discharge :: send a discharge power (+ve) to the power distributor.
4. power_distribution_results :: get a receiver to power distributor responses.

(PSC = Passive Sign Convention)

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Remove tests for multi-user features of PowerDistributor

The multi-user features of PowerDistributor will be removed in a
subsequent commit.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Remove code providing multi-user support in PowerDistributor

Because the PowerDistributor doesn't have a way to resolve conflicts,
it can't be safely used by multiple users. So this feature will never
be used.

Once the PowerArbitrator is available, it will support multiple users
and will be the only component that will communicate with the
PowerDistributor.

This is being done now, in order to make the BatteryPool control
interface to be a bit more efficient and easy to implement.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Add a required `namespace` field to `PowerDistributingActor` requests

This field be used to identify the channel for sending a response
into, in the channel registry.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Update PowerDistributingActor channel interface

Because multi-user support is going away, we can switch to a single
input receiver in the PowerDistributingActor for receiving requests.
So we don't need Bidirectional channels anymore.

We can't have a single output channel, because there would be multiple
request senders (possibly multiple EV Charger Pool instances, with
different component ID sets), and they would need to receive responses
just for the requests they made.

To rephrase, we are removing explicit multi-user support, and
replacing it with implicit multi-user support, which can also go away
once we have the PowerManager.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Update class docstring example for PowerDistributingActor

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Update data pipeline to work with new PowerDistributingActor interface

This commit also removes the `power_distributing_handle` method.  This
will be replaced by methods in the battery_pool that will be added in
a subsequent commit.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Update PowerDistributingActor tests to work with new interface

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Update power distribution example program to work with new interface

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Update power distributor benchmark to run only single-user tests

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

leandro-lucarella-frequenz commented:
I linked a few issues that should be closed by this, I guess this doesn't cover the whole #276, right?  If this closes any other issues, please remember to link them!

shsms commented:
> I linked a few issues that should be closed by this, I guess this doesn't cover the whole https://github.com/frequenz-floss/frequenz-sdk-python/issues/276, right?

Thanks Luca, for #276,  only the bounds part would be pending after this.  I've made an issue for that.

daniel-zullo-frequenz commented:
@sahas-subramanian-frequenz and @leandro-lucarella-frequenz 
based on https://github.com/frequenz-floss/frequenz-sdk-python/pull/365#discussion_r1205082778, do you want to discuss or address the alternative names for `include_broken` for the [request attribute](https://github.com/frequenz-floss/frequenz-sdk-python/blob/v0.x.x/src/frequenz/sdk/actor/power_distributing/request.py#L33) here in this PR? Otherwise I'll create a new issue if this still needs to be discussed/addressed

shsms commented:
I was thinking of `include_broken_batteries`, because that would be slightly more obvious with the `charge/discharge` methods.  Does that make sense?  

But I have no opinion about renaming further,  it turns out it can't be fully explained with just the name,  and we'll just have to use what make sense most of the time (which is include_broken),  and describe it better in the docstrings.

I'll add a further line saying that using this option is still not a guarantee that all/any batteries will get used.

daniel-zullo-frequenz commented:
`include_broken_batteries` is just too long for a flag IMO but as it is an important flag it makes sense to make it explicitly obvious for the user.

> I'll add a further line saying that using this option is still not a guarantee that all/any batteries will get used.

As you gonna add a line, can you also update the documentation for `include_broken_batteries` to something around this:
```
    If set to `True`, the power distribution algorithm will consider all batteries, including the broken ones, when distributing power. 
    In such cases, any remaining power after distributing among the available batteries will be distributed equally among the unavailable (broken) batteries.
    If all batteries in the set are unavailable, the power will be equally distributed among all the unavailable batteries in the request.

    If set to `False`, the power distribution will only take into account the available batteries, excluding any broken ones.
```

leandro-lucarella-frequenz commented:
> `include_broken_batteries` is just too long for a flag IMO but as it is an important flag it makes sense to make it explicitly obvious for the user.

For me it is OK because it is an obscure flag, is not something that we expect users be using all the time. I'm fine with leaving it at `include_broken`, as Sahas said we won't be able to convey the full meaning there, so it should be enough with something that is not clearly confusing or misleading.  I'd rather be as clear as possible though, even if the flag name is on the long side because, again, it should be a rare flag to use anyway.

leandro-lucarella-frequenz commented on src/frequenz/sdk/actor/power_distributing/power_distributing.py:
```diff
@@ -188,21 +178,26 @@ class PowerDistributingActor:
     def __init__(
         self,
-        users_channels: Dict[str, Bidirectional.Handle[Result, Request]],
+        requests_rx: Receiver[Request],
```
I would name this `request_receiver` even if it is a bit longer. I think the `rx`/`tx` idiom might be a bit cryptic for the SDK, at least for the public facing interface.

shsms replied:
done

leandro-lucarella-frequenz commented on src/frequenz/sdk/actor/power_distributing/power_distributing.py:
```diff
@@ -188,21 +178,26 @@ class PowerDistributingActor:
+        self._requests_rx = requests_rx
+        self._channel_registry = channel_registry
         self._wait_for_data_sec = wait_for_data_sec
+        self._result_senders: Dict[str, Sender[Result]] = {}
```
So these are matched to the request namespace, right? Or how do you map which response should be sent to which channel?

Maybe it would be good to clarify adding a docstring to the attribute even if it's private.

shsms replied:
done

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/battery_pool/battery_pool.py:
```diff
@@ -84,15 +90,129 @@ def __init__(  # pylint: disable=too-many-arguments
+        """Set the given power for the batteries in the pool.
+
+        Power values need to follow the Passive Sign Convention (PSC). That is, positive
+        values indicate charge power and negative values indicate discharge power.
```
What about mentioning the existence of `charge()` and `discharge()` here and recommending their use instead (unless you are already working with a power value following PSC)?

shsms replied:
done, also in the charge/discharge methods.

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/battery_pool/battery_pool.py:
```diff
@@ -84,15 +90,129 @@ def __init__(  # pylint: disable=too-many-arguments
+        power: float,
+        *,
+        adjust_power: bool = True,
+        request_timeout_sec: float = 5.0,
```
Shouldn't this be a `timedelta` now?

shsms replied:
updated.

leandro-lucarella-frequenz commented on examples/power_distribution.py:
I wonder if by now we should just remove this example, I guess users should be encouraged to use the battery pool instead of communicating directly to the power distirbuting actor. Maybe not in this PR, but in a near future?

We can leave a small example in the docstring, but the bulk of this example shows how to build the data pipeline more than how to distribute power among a set of batteries.

shsms replied:
True,  we have a number of examples that became more confusing when we updated them to pass pylint as well.  I guess we have a big example cleanup project pending.

leandro-lucarella-frequenz commented on benchmarks/power_distribution/power_distributor.py:
Did you run the benchmarks before and after? Where there any differences?

shsms replied:
It was the same time for the single user tests.

leandro-lucarella-frequenz reviewed (commented):
I keep forgetting why we want to remove the multi-user support for `PowerDistributor`, it might be a good idea to explain the why in the commit message removing it.

Same about the commit adding the `namespace` to the request (I saw it is used in following commits, but when looking at the commit in order it is a bit mysterious).

Besides the small comments, LGTM!

shsms committed:
Rename PowerDistributor request param to `include_broken_batteries`

This makes is a bit more obvious, although longer.  Also updates the
docstring for this field.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms commented:
Renamed to `include_broken_batteries` and updated docstrings.

Updated commit messages explaining removal of multi-user support and introduction of namespaces.

christianparpart commented on src/frequenz/sdk/timeseries/battery_pool/battery_pool.py:
```diff
@@ -84,15 +90,144 @@ def __init__(  # pylint: disable=too-many-arguments
+
+        Args:
+            power: The power to set for the batteries in the pool.
+            adjust_power: If True, the power will be adjusted to fit the power bounds,
+                if necessary. If False, then power requests outside the bounds will be
+                rejected.
```
maybe we should have a better name for `adjust_power`, for example `apply_power_bounds`, because the former does not tell me what this parameter is doing, in fact, I was scratching my face and wondering the whole purpose of this function just by looking at that boolean in that context. :-)

With a better name, hopefully (like proposed), it should be more clear what impact this boolean parameter will have.

Same applies to the other `adjust_power` occurrences below.

leandro-lucarella-frequenz replied:
Another alternative could be `adjust_to_bounds`. at least for me it is not super clear what applying a bound means.

shsms replied:
I think it is impossible to encode everything the flag does into its name,  we will have to read the docs,  because it is not trivial and it needs info about
1. when/why is it required - when requests ~exceed the bounds~ are out of bounds
2. what does it do when true - adjust power to fit within the bounds and forward request
3. what does it do when false - reject request,  not forward without adjusting.

And `apply_power_bounds` always happens.  What to do after that is the question - adjust the power value to fit the bounds, or reject the request.

shsms replied:
But I agree, `adjust_to_bounds` is nicer.  If there's code that says:

```python
await battery_pool.charge(2000, adjust_to_bounds=True)
```
maybe it is obvious to first time users without reading the docs.  Hopefully,  in most scenarios they want our default value (True) anyway and won't have to deal with it.

llucax replied:
> I think it is impossible to encode everything the flag does into its name, we will have to read the docs, because it is not trivial and it needs info about

Yeah, I was about to say the same thing, but in this case I also felt that there was room for improvement :)

We need to strike a balance. And like `sop`, if something is super hard to make clear in a *indetifier*, I think it is better to have it obscure to force people to look at the docs instead of having something more meaningful but misleading.
